File: fn/src/main/java/org/sqlbuilder/fn/joins/FE_FERequired.java
package org.sqlbuilder.fn.joins;

import org.sqlbuilder.common.Insertable;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Set;

import edu.berkeley.icsi.framenet.InternalFrameRelationFEType;

public class FE_FERequired extends Pair<Integer, Integer> implements Insertable
{
	public static final Set<FE_FERequired> SET = new HashSet<>();

	// C O N S T R U C T O R

	@SuppressWarnings("UnusedReturnValue")
	public static FE_FERequired make(final int fe, final InternalFrameRelationFEType fe2)
	{
		var ff = new FE_FERequired(fe, fe2.getID());
		SET.add(ff);
		return ff;
	}

	private FE_FERequired(final int feid, final int feid2)
	{
		super(feid, feid2);
	}

	// O R D E R

	public static final Comparator<FE_FERequired> COMPARATOR = Comparator.comparing(FE_FERequired::getFirst).thenComparing(FE_FERequired::getSecond);

	// I N S E R T

	@Override
	public String dataRow()
	{
		return String.format("%d,%d", first, second);
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[FE-FEreq feid=%s feid2=%s]", first, second);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/joins/FE_FERequired.kt
package org.sqlbuilder.fn.joins

import edu.berkeley.icsi.framenet.InternalFrameRelationFEType
import org.sqlbuilder.common.Insertable
import java.util.*

data class FE_FERequired(
    val feid: Int,
    val feid2: Int,
) : Insertable {

    // I N S E R T

    override fun dataRow(): String {
        return "$feid,$feid2"
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as FE_FERequired
        return feid == that.feid && feid2 == that.feid2
    }

    override fun hashCode(): Int {
        return Objects.hash(feid, feid2)
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[FE-reqFE feid=$feid feid2=$feid2]"
    }

    companion object {

        @JvmField
        val COMPARATOR: Comparator<FE_FERequired> = Comparator
            .comparing<FE_FERequired, Int> { it.feid }
            .thenComparing<Int> { it.feid2 }

        @JvmField
        val SET = HashSet<FE_FERequired>()

         @JvmStatic
        fun make(fe: Int, fe2: InternalFrameRelationFEType): FE_FERequired {
            val ff = FE_FERequired(fe, fe2.getID())
            SET.add(ff)
            return ff
        }
    }
}
